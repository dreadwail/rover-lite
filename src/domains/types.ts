import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
