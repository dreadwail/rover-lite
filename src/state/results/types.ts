export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
