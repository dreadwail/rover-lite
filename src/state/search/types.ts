import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
