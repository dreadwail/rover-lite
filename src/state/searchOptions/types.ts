export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
