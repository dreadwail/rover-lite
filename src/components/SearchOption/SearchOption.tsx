import React, { PureComponent } from 'react';

import {
  isBooleanOption,
  isChoiceOption,
  isDateOption,
  isIntegerOption,
  isMultipleChoiceOption,
  Option,
} from '../../state/searchOptions/types';
import SearchOptionBoolean from '../SearchOptionBoolean';
import SearchOptionChoice from '../SearchOptionChoice';
import SearchOptionDate from '../SearchOptionDate';
import SearchOptionInteger from '../SearchOptionInteger';
import SearchOptionMultipleChoice from '../SearchOptionMultipleChoice';

export interface SearchOptionProps {
  readonly optionKey: string;
  readonly option: Option;
  readonly value?: any;
  onChange(value: any): void;
}

export default class SearchOption extends PureComponent<SearchOptionProps> {
  public render() {
    const { optionKey, option, value, onChange } = this.props;
    if (isDateOption(option)) {
      return (
        <SearchOptionDate optionKey={optionKey} option={option} value={value} onChange={onChange} />
      );
    }
    if (isIntegerOption(option)) {
      return (
        <SearchOptionInteger
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    if (isBooleanOption(option)) {
      return (
        <SearchOptionBoolean
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    if (isMultipleChoiceOption(option)) {
      return (
        <SearchOptionMultipleChoice
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    if (isChoiceOption(option)) {
      return (
        <SearchOptionChoice
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    return null;
  }
}
