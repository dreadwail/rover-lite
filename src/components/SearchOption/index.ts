export * from './SearchOption';
export { default } from './SearchOption';
