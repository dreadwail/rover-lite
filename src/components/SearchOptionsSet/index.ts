export * from './SearchOptionsSet';
export { default } from './SearchOptionsSet';
