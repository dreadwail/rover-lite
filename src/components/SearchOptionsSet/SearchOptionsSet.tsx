import React, { FC } from 'react';

import { SearchInputs } from '../../state/search/types';
import { Options } from '../../state/searchOptions/types';
import SearchOption from '../SearchOption';

export interface SearchOptionsSetProps {
  readonly options: Options;
  readonly inputs: SearchInputs;
  onSearchFieldChanged(key: string, value: any): void;
}

const SearchOptionsSet: FC<SearchOptionsSetProps> = ({ options, inputs, onSearchFieldChanged }) => {
  const keys = Object.keys(options);
  if (keys.length === 0) {
    return null;
  }
  return (
    <>
      {keys.map(optionKey => {
        const option = options[optionKey];
        const value = inputs[optionKey];
        const onChange = (newValue: any) => {
          onSearchFieldChanged(optionKey, newValue);
        };
        return (
          <SearchOption
            key={optionKey}
            optionKey={optionKey}
            option={option}
            value={value}
            onChange={onChange}
          />
        );
      })}
    </>
  );
};

export default SearchOptionsSet;
