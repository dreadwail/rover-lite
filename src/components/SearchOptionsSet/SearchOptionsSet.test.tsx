import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import { IntegerOption } from '../../state/searchOptions/types';
import SearchOption from '../SearchOption';

import SearchOptionsSet, { SearchOptionsSetProps } from './SearchOptionsSet';

describe('SearchOptionsSet', () => {
  let rendered: ShallowWrapper<SearchOptionsSetProps>;
  let props: SearchOptionsSetProps;

  describe('with no options', () => {
    beforeEach(() => {
      props = {
        options: {},
        inputs: {},
        onSearchFieldChanged: jest.fn().mockName('onSearchFieldChanged'),
      };
      rendered = shallow(<SearchOptionsSet {...props} />);
    });

    it('renders nothing', () => {
      expect(rendered.isEmptyRender()).toBe(true);
    });
  });

  describe('with options', () => {
    beforeEach(() => {
      const option: IntegerOption = {
        type: 'integer',
        minValue: 0,
        maxValue: 50,
        required: false,
        label: 'test-label',
        category: {
          categoryId: 'test-category-id',
          rank: 0,
          displayName: 'Test Category',
        },
        rank: 0,
      };
      props = {
        options: { test_key: option },
        inputs: { test_key: 42 },
        onSearchFieldChanged: jest.fn().mockName('onSearchFieldChanged'),
      };
      rendered = shallow(<SearchOptionsSet {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });

    describe('when an option is changed', () => {
      const newValue: number = 42;

      beforeEach(() => {
        const searchOption = rendered.find(SearchOption);
        const onChange = searchOption.prop('onChange');
        onChange(newValue);
      });

      it('calls the onSearchFieldChanged prop with the key and new value', () => {
        expect(props.onSearchFieldChanged).toHaveBeenCalledWith('test_key', newValue);
      });
    });
  });
});
