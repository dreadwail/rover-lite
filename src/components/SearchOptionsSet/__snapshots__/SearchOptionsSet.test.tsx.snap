// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchOptionsSet with options renders correctly 1`] = `
<Fragment>
  <SearchOption
    key="test_key"
    onChange={[Function]}
    option={
      Object {
        "category": Object {
          "categoryId": "test-category-id",
          "displayName": "Test Category",
          "rank": 0,
        },
        "label": "test-label",
        "maxValue": 50,
        "minValue": 0,
        "rank": 0,
        "required": false,
        "type": "integer",
      }
    }
    optionKey="test_key"
    value={42}
  />
</Fragment>
`;
